ouroboros-network/src/Ouroboros/Network/Protocol/TipSample/Type.hs

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE EmptyCase         #-}
+{-# LANGUAGE PolyKinds         #-}
+
+module Ouroboros.Network.Protocol.TipSample.Type where
+
+import Network.TypedProtocol.Core
+
+
+data TipRequestKind where
+    BlockUntilSlot      :: TipRequestKind
+    BlockUntilTipChange :: TipRequestKind
+
+data TokTipRequest (tipRequestKind :: TipRequestKind) where
+    TokBlockUntilSlot      :: TokTipRequest BlockUntilSlot
+    TokBlockUntilTipChange :: TokTipRequest BlockUntilTipChange
+
+-- | `tip-sample` protocols, desigined for established peers to sample tip from
+-- upstream peers.
+--
+data TipSample tip slotNo where
+    StStart      :: TipSample tip slotNo
+    StTipRequest :: TipSample tip slotNo
+    StIdle       :: TipSample tip slotNo
+    StBusy       :: TipRequestKind -> TipSample tip slotNo
+    StDone       :: TipSample tip slotNo
+
+
+instance Protocol (TipSample tip slotNo) where
+    data Message (TipSample tip slotNo) from to where
+
+      -- | Request current tip.  It must not block.
+      --
+      MsgGetCurrentTip :: Message (TipSample tip slotNo)
+                                  StStart StTipRequest
+
+      -- | Server response with its current tip.
+      --
+      MsgCurrentTip :: tip -> Message (TipSample tip slotNo)
+                                       StTipRequest StIdle
+
+
+      -- | Get first tip after given slot.  It can block.
+      --
+      MsgGetTipAfterSlotNo :: slotNo -> Message (TipSample tip slotNo)
+                                                StIdle (StBusy BlockUntilSlot)
```

Comment:
`MsgGetTipAfterSlotNo` does not see forks, it just waits until we pass a given `SlotNo`:  it can happen that a server replies and before even it gets the next request, a fork happens and now it will block until the next given slot is passed.  I think this is ok. The measurements will include how well the node is placed in the network (if a node needs to follow less forks to get to a given slot, it have a greater chance to serve us the final block slightly earlier).